File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/transactions/presenter/TransactionDetailPresenter.java
package org.mifos.mobilewallet.mifospay.transactions.presenter;

import org.mifos.mobilewallet.core.base.UseCaseHandler;
import org.mifos.mobilewallet.core.domain.model.Transaction;
import org.mifos.mobilewallet.core.domain.usecase.account.FetchTransactionReceipt;
import org.mifos.mobilewallet.mifospay.base.BaseView;
import org.mifos.mobilewallet.mifospay.transactions.TransactionsContract;

import java.util.ArrayList;

import javax.inject.Inject;

/**
 * @author ankur
 * @since 06/June/2018
 */

public class TransactionDetailPresenter implements TransactionsContract.TransactionDetailPresenter {

    private final UseCaseHandler mUseCaseHandler;
    @Inject
    FetchTransactionReceipt mFetchTransactionReceiptUseCase;
    private TransactionsContract.TransactionDetailView mTransactionDetailView;

    @Inject
    public TransactionDetailPresenter(UseCaseHandler useCaseHandler) {
        mUseCaseHandler = useCaseHandler;
    }

    @Override
    public void attachView(BaseView baseView) {
        mTransactionDetailView = (TransactionsContract.TransactionDetailView) baseView;
        mTransactionDetailView.setPresenter(this);
    }

    /**
     * Method to return specific transactions from list of transactions.
     * @param transactions the list of transactions
     * @param secondAccountNumber the Account Number
     * @return list of specific transactions
     */
    @Override
    public ArrayList<Transaction> getSpecificTransactions(ArrayList<Transaction> transactions,
            String secondAccountNumber) {
        ArrayList<Transaction> specificTransactions = new ArrayList<>();
        for (Transaction transaction : transactions) {
            if (transaction.getTransferDetail() != null
                    && (transaction.getTransferDetail().getFromAccount().getAccountNo().equals(
                    secondAccountNumber)
                    || transaction.getTransferDetail().getToAccount().getAccountNo().equals(
                    secondAccountNumber))) {

                specificTransactions.add(transaction);
            }
        }
        return specificTransactions;
    }
}

File: mifospay/src/main/java/org/mifospay/history/transaction_detail/presenter/TransactionDetailPresenter.kt
package org.mifospay.history.transaction_detail.presenter

import org.mifospay.core.data.base.UseCase.UseCaseCallback
import org.mifospay.core.data.base.UseCaseHandler
import org.mifospay.core.data.domain.usecase.account.FetchAccountTransfer
import org.mifospay.base.BaseView
import org.mifospay.history.HistoryContract
import org.mifospay.history.HistoryContract.TransactionDetailView
import org.mifospay.common.Constants
import javax.inject.Inject

/**
 * Created by ankur on 06/June/2018
 */
class TransactionDetailPresenter @Inject constructor(
    private val mUseCaseHandler: UseCaseHandler,
    private val mFetchAccountTransferUseCase: FetchAccountTransfer
) : HistoryContract.TransactionDetailPresenter {

    private var mTransactionDetailView: TransactionDetailView? = null

    override fun attachView(baseView: BaseView<*>?) {
        mTransactionDetailView = baseView as TransactionDetailView?
        mTransactionDetailView!!.setPresenter(this)
    }

    override fun getTransferDetail(transferId: Long) {
        mTransactionDetailView!!.showProgressBar()
        mUseCaseHandler.execute(mFetchAccountTransferUseCase,
            FetchAccountTransfer.RequestValues(transferId),
            object : UseCaseCallback<FetchAccountTransfer.ResponseValue?> {

                override fun onSuccess(response: FetchAccountTransfer.ResponseValue?) {
                    mTransactionDetailView!!.hideProgressBar()
                    mTransactionDetailView!!.showTransferDetail(response?.transferDetail)
                }

                override fun onError(message: String) {
                    mTransactionDetailView!!.hideProgressBar()
                    mTransactionDetailView!!.showToast(
                        Constants.ERROR_FETCHING_TRANSACTION_DETAILS
                    )
                }
            })
    }
}